README.md
# hdhr-ac4 v1.5.0

## Description

>This project aims to emulate an HDHomerun tuner that supplies ATSC 3.0 programs with AC3 audio.

## About

>This project started as a way for me to support ATSC 3.0 programs on my home netowrk. I already have an HDHR5-4k tuner and thought the transition would be straight forward when ATSC 3.0 programs went live in Atlanta. The signal and video quality were much better, but the AC4 audio every program was using was supported by nothing I owned.

>This application pretends to be an HDHomerun device that offers ATSC 3.0 programs. When a request comes in to tune one of the programs, it forwards that request to a real HDHomerun device and passes the resulting stream back to the original requester after converting the AC4 audio stream to AC3 audio.

## Why?

>This allows my HDHR5-4K to serve ATSC 3.0 programs to my Emby server(now native support?) and PLEX server with Live TV and DVR functionality. Converting audio to AC3 before the server receives it allows everything to work as it always has with HDHomerun tuners.

## Configuration 
>Set in docker environent
>>- HDHR_IP - REQUIRED. The IP address of your real HDHomerun device
>>- HOST_IP - REQUIRED. The IP address where hdhr-ac4 is running
>>- DEVICEID_SWAP - OPTIONAL (1 or 0, default 0) If the Device of the HDHomerun you are connecting to should be reversed. This ensures you have a unique DeviceID to prevent any collision with your physical HDHomerun device.

>### hd_home_run.py
>>Contains the ffmpeg command:
```
[
    "/bin/usr/ffmpeg",
    "-nostats",
    "-hide_banner",
    "-loglevel",
    "warning",
    "-i",
    "pipe:",
    "-c:a",
    "ac3",
    "-c:v",
    "copy",
    "-f",
    "mpegts",
    "-",
]
```
>>This should not need to be changed, but can be modified if the ATSC 3.0 streams needs specifi ffmpeg handling.

## Build Docker Container
>Download the latest Emby installer from https://emby.media/linux-server.html and save it in the docker build directory. Update Dockerfile to point to the file. Currently using emby-server-deb_4.8.0.21_amd64.deb
>Example Dockerfile build command:

>`docker build -f Dockerfile -t hdhr-ac4 .`

>This build uses ffmpeg binary from Emby. The Emby team has a custom version of ffmpeg tha has several improvements over the original ffmpeg branch. 
>Version 1.0.0 of this docker build created the custom ffmpeg from scratch which took much longer and is a little dated at this point. 
>A quick explanation for the docker build that is based on ubuntu 20:
>1. The ffmpeg container extracts Emby installer
>2. The final container copies in ffmpeg binaries. Python is installed and a few python modules added. The 2 python files are copied over and the launch command is set.

## Run Docker container
>Example container run command:

>`docker run -p 80:80 -p 5004:5004 -e "HDHR_IP=10.1.1.2" -e "HOST_IP=10.0.0.100" -e "DEVICEID_SWAP=1" hdhr-ac4`

>The HDHomerun API being implemented is here: https://info.hdhomerun.com/info/http_api 

>The container runs HTTP servers on port 80 and 5004 and supports the following HTTP requests:
>>- http://HOST:80/ - returns info and version about hdhr-ac4 docker
>>- http://HOST:80/discover.json - returns the same response as the real HDHomerun device, but substitutes the hdhr-ac4 IP address in the appropriate locations.
>>- http://HOST:80/lineup.json - queries the real HDHomerun device for its service list and then does the following:
>>>1. replaces all the IP addresses with the hdhr-ac4 IP address
>>>2. replaces the ATSC3=1 entry with AudioCodec=AC3
>>>3. strips out all non ATSC3 programs
>>- http://HOST:5004/auto/{program} - forwards the request to the real HDHomerun device, and then pipes the stream through ffmpeg and then back to the orignal requester
>>- http://HOST:80/lineup_status.json - returns the same response as the real HDHomerun device

## Notes
>- Please report issues or send pull requests here: https://github.com/johnb-7/hdhr-ac4
>- Channel changes are a little slower due to the extra step
>- VLC was used a lot in the early testing. Example URL for host 192.168.1.1 with program 111.1: http://192.168.1.1:5004/auto/v111.1
>- The development container (Dockerfile-dev) is very similar but does not automatically launch the application and can be mounted using your editor of choice for debugging the Python application or ffmpeg. It also installs some extra software specifically for development.
>- Many thanks to the Emby team for the ffmpeg with working AC4. Since its open source software, there should not be any problems here.

## TODO
>-Make official release on docker hub

## License
>This project is release under the Apache 2.0 license


CHANGELOG.md
>1.5.0
>- major docker build update uses Emby pre-built ffmpeg binaries with excellent AC4 support
>- removed async argument no longer needed with Emby ffmpeg
>- updated modules to latest versions
>- added multiple workers to uvicorn for faster tune time and better handling of multiple tunes
>- configuration is now done through docker environment variables
>- added info API that also reports version
>- documentation update


>1.0.0
>- initial release
>- hdhr proxy that supplies ATSC 3.0 channels and supports streaming them with audio re-encoded to AC3
>- docker build does full compile of ffmprg ac4 branch
>- special thanks to gunsuka for adding lineup_status.json API and DeviceID_swap
>- tested on Emby and Plex


hd_home_run.py
"""Class to represent a HD Home Run device on the network"""
import requests
from fastapi import Request, HTTPException
from fastapi.responses import StreamingResponse
from subprocess import Popen, PIPE
from requests import Response
from threading import Thread
from time import sleep


def stream_requests_to_ffmpeg(response: Response, ffmpeg: Popen):
    try:
        for data in response.iter_content(chunk_size=1024 * 128):
            if ffmpeg.poll() != None:
                break
            written = ffmpeg.stdin.write(data)
            if written <= 0:
                break
            # sleep(0.01)
    except:
        pass


async def stream_ffmpeg_to_response(ffmpeg: Popen, feeder: Thread, request: Request):
    while 1:
        try:
            data = ffmpeg.stdout.read(1024 * 128)
            if (
                len(data) <= 0
                or not feeder.is_alive()
                or await request.is_disconnected()
            ):
                break
            yield data
            # sleep(0.01)
        except:
            break
    await request.close()
    ffmpeg.kill()


class HdHomeRun:
    _base_url: str

    def __init__(self, ip: str) -> None:
        """Init the class with the device IP"""
        self._base_url = "http://" + ip

    def discover(self) -> str:
        """Gets the hd home run info"""
        return requests.get(self._base_url + "/discover.json").text

    def lineup(self) -> str:
        """Gets the hd home run channel list"""
        return requests.get(self._base_url + "/lineup.json").text

    def lineup_status(self) -> str:
        """Gets the hd home run status"""
        return requests.get(self._base_url + "/lineup_status.json").text

    def tune(self, channel: str, stream_out: Request) -> None:
        """Streams channel from hdhr until device disconnects"""
        stream_in = requests.get(self._base_url + ":5004/auto/" + channel, stream=True)
        if stream_in.status_code == 200:
            ffmpeg = Popen(
                [
                    "/usr/bin/ffmpeg",
                    "-nostats",
                    "-hide_banner",
                    "-loglevel",
                    "warning",
                    "-i",
                    "pipe:",
                    "-c:a",
                    "ac3",
                    "-c:v",
                    "copy",
                    "-f",
                    "mpegts",
                    "-",
                ],
                stdout=PIPE,
                stdin=PIPE,
            )
            feeder = Thread(target=stream_requests_to_ffmpeg, args=(stream_in, ffmpeg))
            feeder.start()
            return StreamingResponse(
                stream_ffmpeg_to_response(ffmpeg, feeder, stream_out),
                headers=stream_in.headers,
            )
        else:
            stream_in.close()
            del stream_in.headers["Content-Length"]
            raise HTTPException(
                status_code=stream_in.status_code, headers=stream_in.headers
            )


main.py
"""main file tha launches fastapi and created hdhr instance"""
from typing import Any, List
from fastapi import FastAPI, Request
from fastapi.exceptions import HTTPException
from fastapi.responses import StreamingResponse
from hd_home_run import HdHomeRun
from threading import Thread
import uvicorn
from pydantic import BaseModel
import json
import re
import os
import sys

# These must be declared in the docker or shell environment
HDHR_IP = os.getenv("HDHR_IP")
HOST_IP = os.getenv("HOST_IP")
if HDHR_IP is None or HOST_IP is None:
    print("ERROR: missing HDHR_IP or HOSTIP")
    sys.exit(1)

# These config options are optionl
# Set to 1 to reverse the DeviceID of the original HDHR.
# This is needed for some systems like PLEX that track the DeviceID
try:
    DeviceID_swap = int(os.getenv("DEVICEID_SWAP"))
except:
    print("WARN: Default DeviceID_swap = 0")
    DeviceID_swap = 0

app = FastAPI()
tune = FastAPI()
hdhr_instance = HdHomeRun(HDHR_IP)


@app.get("/")
async def get_info():
    return {
        "application": "hdhr-ac4",
        "version": "1.5.0",
        "website": "https://github.com/johnb-7/hdhr-ac4",
    }


@app.get("/discover.json")
async def get_discover():
    original = hdhr_instance.discover()
    modified = original.replace(HDHR_IP, HOST_IP)

    if DeviceID_swap:
        DID_search = re.search(r'"DeviceID":"([A-F0-9]+)"', modified)
        if DID_search:
            modified = re.sub(
                r'"DeviceID":"([A-F0-9]+)"',
                r'"DeviceID":"' + DID_search.group(1)[::-1] + '"',
                modified,
            )
    return json.loads(modified)


@app.get("/lineup.json")
async def get_lineup():
    original_txt = hdhr_instance.lineup()
    modified_txt = original_txt.replace(HDHR_IP, HOST_IP).replace(
        '"ATSC3":1', '"AudioCodec":"AC3"'
    )
    original_json = json.loads(modified_txt)
    modified_json = []
    for entry in original_json:
        if "VideoCodec" in entry and entry["VideoCodec"] == "HEVC":
            print(entry)
            modified_json.append(entry)
    return modified_json


@app.get("/lineup_status.json")
async def get_lineup_status():
    original_json = hdhr_instance.lineup_status()
    return json.loads(original_json)


@tune.get("/auto/{channel}")
async def in_channel(channel: str, request: Request) -> Any:
    return hdhr_instance.tune(channel, request)


if __name__ == "__main__":
    # only for dev, prod runs through uvicorn command line
    app_thread = Thread(
        target=uvicorn.run, kwargs={"app": app, "port": 80, "host": "0.0.0.0"}
    )
    tune_thread = Thread(
        target=uvicorn.run, kwargs={"app": tune, "port": 5004, "host": "0.0.0.0"}
    )
    app_thread.start()
    tune_thread.start()
    app_thread.join()
    tune_thread.join()
